### ISSUE TITLE
MacroServer hangs during scan

### ISSUE THREAD

**Issue by daneos** [sardana-org/sardana#736]
This is the issue we have been experiencing at Solaris for a long time, and we finally have some insight of what's happening.

From time to time, MacroServer hangs during scan. It seems totally random, but happens everytime if you perform multiple scans, or your scans are really long. The problem is reproducible both on CentOS 7 and Debian 9, whenever using distribution's Tango packages or custom built ones. The catch is, to reproduce the problem you need to control a real hardware motor for your scan, for some reason on DummyMotor this doesn't happen (or at least scans would need to be significantly longer). I used IcePAP for testing, not sure if other motor controllers will trigger this problem too.

What we have found is that immediately before hanging, after getting values from expchans, but before starting movement to the next point, MacroServer tries to subscribe to mntgrp `Moveable` attribute. This fails due to polling not being started on that attribute, and indeed it's not enabled there by default. You can see that on the screenshots below.
![screenshot1](https://user-images.githubusercontent.com/2900361/39055911-a5809c22-44b5-11e8-9233-a4dc92ef4f27.png)
![screenshot2](https://user-images.githubusercontent.com/2900361/39055934-b26b493c-44b5-11e8-94a2-a85f1d0fce5c.png)
This however raises several questions:
* What causes this subscription attempt? All scan points before were fine without it, why it's needed in the middle of the scan?
* If this subsciption is needed, shouldn't the polling be enabled on mntgrp creation?
* The code for Moveable attribute is more or less the same as for other mntgrp attributes that can push events from code. Is this attribute different, or should it be marked as pushing events from code?

We also tried to pinpoint where the subscription is made and why, but it seems that code executed when Sardana hangs has nothing to do with this. It stops always in the same place in code, so the subscription probably isn't triggered by some other thread.

Enabling polling on the `Moveable` attribute solves the problem, but I think it should rather be considered a workaround and not a fix.

**Comment by reszelaz** [comment]
Thanks for this investigation!

There was an inconsistency in the code and this attribute should be marked as pushing change events from the code. The client side (taurus extension that resides in the MacroServer) relies on this events - it adds a listener (see `TangoAttributeEG` class). What happens is that taurus first tries to subscribe with the stateless flag set to False and if this fails (in this case it clearly fails cause there is no polling nor the events are supposed to be pushed from the code), enables its own (taurus) polling and subscribes with the stateless set to True. Then I think that Tango will periodically attemp to subscribe to the events. I imagine that this is the subscription that causes problems to you. This taurus polling was active during the whole scan.

The same applies to Synchronization and LatencyTime attributes.

I've proposed a #738 to fix this issue, could you review it? I've set these attributes to detect the event change. I don't know why the rest of the MeasurementGroup attributes are marked as non detect the change. I will create another issue to clarify this.

What is strange is why this hangs the MacroServer? At the end what is happening is that the client (which in this case is another device server) is subscribed to the MeasurementGroup attributes with the stateless flag set to True and it reads the attribute with 3s polling period. I think it could be worth to try to reproduce it with device servers independt to Sardana and eventually, if the issue can be reproduced, report it to the Tango community.

The good thing is that thanks to this discovery and the fix we will eliminate useless polling done by the Macroserver to these three MeasurementGroup attribute during the scans.
